juspay/hyperswitch
# feat(dynamic_routing): analytics improvement using separate postgres table

## Type of Change



## Description

Previously we were pushing the Dynamic routing metrics to Prometheus, but due to high cardinality of some values(payment_id) we were not able to add that in our attributes, hence we are pushing this after creating a new table in our Database (dynamic_routing_stats).

The further enhancement will be done in this track:
https://github.com/juspay/hyperswitch/issues/6756

### Additional Changes





## Motivation and Context

Required to get the payment_id for our metrics purpose.

## How did you test it?

Can be tested in the following ways:
1. enable success_based_routing(feature = metrics) for merchant.
```
curl --location --request POST 'http://localhost:8080/account/m_id/business_profile/pro_id/dynamic_routing/success_based/toggle?enable=metrics' \
--header 'api-key:api_key'
```
2. Make a payment afterwards.
3. Check the `dynamic_routing_stats` table for the specific value associated with the `payment_id`.
<img width="645" alt="Screenshot 2024-12-06 at 6 31 56 PM" src="https://github.com/user-attachments/assets/8317e26b-347d-489a-a8fa-661ab692b157">


## Migration
```
CREATE TYPE "SuccessBasedRoutingConclusiveState" AS ENUM(
  'true_positive',
  'false_positive',
  'true_negative',
  'false_negative'
);

CREATE TABLE IF NOT EXISTS dynamic_routing_stats (
    payment_id VARCHAR(64) NOT NULL,
    attempt_id VARCHAR(64) NOT NULL,
    merchant_id VARCHAR(64) NOT NULL,
    profile_id VARCHAR(64) NOT NULL,
    amount BIGINT NOT NULL,
    success_based_routing_connector VARCHAR(64) NOT NULL,
    payment_connector VARCHAR(64) NOT NULL,
    currency "Currency",
    payment_method VARCHAR(64),
    capture_method "CaptureMethod",
    authentication_type "AuthenticationType",
    payment_status "AttemptStatus" NOT NULL,
    conclusive_classification "SuccessBasedRoutingConclusiveState" NOT NULL,
    created_at TIMESTAMP NOT NULL,
    PRIMARY KEY(attempt_id, merchant_id)
);
CREATE INDEX profile_id_index ON dynamic_routing_stats (profile_id);

```

## Checklist
